In [1]:
%load_ext sql
%sql postgresql://appdev@data:5432/appdev

'Connected: appdev@appdev'

# 1. On the table circuits report:
* What type of indices exists for the table and why they are of that type (and not some other type)
* The amount of space each index takes up

To do this assignmenet i have used the following commands

```
1. \d+ f1db.circuits
2. \di+ idx_17102_primary
3. \di+ idx_17102_url
4. \di+ circuits_position_idx
```

![alt text](https://github.com/philliphb/materialised-views/circuits index.png "Circuits Index")

As we can see from the picture above there are three indices in the circuits table. There are the 
primary (circuitid), url(url) and position(position). Circuitid is a btree, url is also a btree and position is a gist.

# 2. We are talent scouts looking to win over some of the best new drivers there are. But we don't want them too old! Write a query that finds the winner of all the races, but only if they are younger than 38 years. The query should give return the date, driver surname, driver age, track time in milliseconds, race name and circuit name for all races.

In [29]:
%%sql
SELECT drivers.surname AS driver_surname, drivers.dob AS driver_age, races.name AS race_name, races.date AS race_date, circuits.name AS circuits_name, results.milliseconds AS milliseconds  FROM results 
JOIN drivers ON drivers.driverid = results.driverid AND drivers.dob > '1980-01-01' AND results.position = 1
JOIN races ON races.raceid = results.raceid
JOIN circuits ON circuits.circuitid = races.circuitid
ORDER BY drivers.dob DESC;

199 rows affected.


[('Verstappen', datetime.date(1997, 9, 30), 'Malaysian Grand Prix', datetime.date(2017, 10, 1), 'Sepang International Circuit', 5401290),
 ('Verstappen', datetime.date(1997, 9, 30), 'Spanish Grand Prix', datetime.date(2016, 5, 15), 'Circuit de Barcelona-Catalunya', 6100017),
 ('Bottas', datetime.date(1989, 8, 28), 'Austrian Grand Prix', datetime.date(2017, 7, 9), 'Red Bull Ring', 4908523),
 ('Bottas', datetime.date(1989, 8, 28), 'Russian Grand Prix', datetime.date(2017, 4, 30), 'Sochi Autodrom', 5288743),
 ('Ricciardo', datetime.date(1989, 7, 1), 'Azerbaijan Grand Prix', datetime.date(2017, 6, 25), 'Baku City Circuit', 7435573),
 ('Ricciardo', datetime.date(1989, 7, 1), 'Belgian Grand Prix', datetime.date(2014, 8, 24), 'Circuit de Spa-Francorchamps', 5076556),
 ('Ricciardo', datetime.date(1989, 7, 1), 'Canadian Grand Prix', datetime.date(2014, 6, 8), 'Circuit Gilles Villeneuve', 5952830),
 ('Ricciardo', datetime.date(1989, 7, 1), 'Hungarian Grand Prix', datetime.date(2014, 7, 27), 'Hungaroring', 6785058),
 ('Ricciardo', datetime.date(1989, 7, 1), 'Malaysian Grand Prix', datetime.date(2016, 10, 2), 'Sepang International Circuit', 5832776),
 ('Vettel', datetime.date(1987, 7, 3), 'Korean Grand Prix', datetime.date(2012, 10, 14), 'Korean International Circuit', 5788651),
 ('Vettel', datetime.date(1987, 7, 3), 'Japanese Grand Prix', datetime.date(2012, 10, 7), 'Suzuka Circuit', 5336242),
 ('Vettel', datetime.date(1987, 7, 3), 'Bahrain Grand Prix', datetime.date(2012, 4, 22), 'Bahrain International Circuit', 5710990),
 ('Vettel', datetime.date(1987, 7, 3), 'Singapore Grand Prix', datetime.date(2012, 9, 23), 'Marina Bay Street Circuit', 7226144),
 ('Vettel', datetime.date(1987, 7, 3), 'Hungarian Grand Prix', datetime.date(2017, 7, 30), 'Hungaroring', 5986713),
 ('Vettel', datetime.date(1987, 7, 3), 'Italian Grand Prix', datetime.date(2013, 9, 8), 'Autodromo Nazionale di Monza', 4713352),
 ('Vettel', datetime.date(1987, 7, 3), 'Bahrain Grand Prix', datetime.date(2013, 4, 21), 'Bahrain International Circuit', 5760498),
 ('Vettel', datetime.date(1987, 7, 3), 'Malaysian Grand Prix', datetime.date(2013, 3, 24), 'Sepang International Circuit', 5936681),
 ('Vettel', datetime.date(1987, 7, 3), 'Belgian Grand Prix', datetime.date(2011, 8, 28), 'Circuit de Spa-Francorchamps', 5204893),
 ('Vettel', datetime.date(1987, 7, 3), 'Italian Grand Prix', datetime.date(2011, 9, 11), 'Autodromo Nazionale di Monza', 4846172),
 ('Vettel', datetime.date(1987, 7, 3), 'Singapore Grand Prix', datetime.date(2011, 9, 25), 'Marina Bay Street Circuit', 7146757),
 ('Vettel', datetime.date(1987, 7, 3), 'Indian Grand Prix', datetime.date(2012, 10, 28), 'Buddh International Circuit', 5470744),
 ('Vettel', datetime.date(1987, 7, 3), 'Korean Grand Prix', datetime.date(2011, 10, 16), 'Korean International Circuit', 5881994),
 ('Vettel', datetime.date(1987, 7, 3), 'Indian Grand Prix', datetime.date(2011, 10, 30), 'Buddh International Circuit', 5435002),
 ('Vettel', datetime.date(1987, 7, 3), 'Indian Grand Prix', datetime.date(2013, 10, 27), 'Buddh International Circuit', 5472187),
 ('Vettel', datetime.date(1987, 7, 3), 'Japanese Grand Prix', datetime.date(2013, 10, 13), 'Suzuka Circuit', 5209301),
 ('Vettel', datetime.date(1987, 7, 3), 'British Grand Prix', datetime.date(2009, 6, 21), 'Silverstone Circuit', 4969328),
 ('Vettel', datetime.date(1987, 7, 3), 'Bahrain Grand Prix', datetime.date(2017, 4, 16), 'Bahrain International Circuit', 5633374),
 ('Vettel', datetime.date(1987, 7, 3), 'Japanese Grand Prix', datetime.date(2009, 10, 4), 'Suzuka Circuit', 5300443),
 ('Vettel', datetime.date(1987, 7, 3), 'Abu Dhabi Grand Prix', datetime.date(2009, 11, 1), 'Yas Marina Circuit', 5643414),
 ('Vettel', datetime.date(1987, 7, 3), 'Korean Grand Prix', datetime.date(2013, 10, 6), 'Korean International Circuit', 6193701),
 ('Vettel', datetime.date(1987, 7, 3), 'Singapore Grand Prix', datetime.date(2013, 9, 22), 'Marina Bay Street Circuit', 7153132),
 ('Vettel', datetime.date(1987, 7,

# 3. Describe the query using EXPLAIN ANALYZE with at least 5 lines of text. Answer at least the following:
* How many calls are you making?
* How long does it take to perform the query?

In [36]:
%%sql EXPLAIN ANALYZE
SELECT drivers.surname AS driver_surname, drivers.dob AS driver_age, races.name AS race_name, races.date AS race_date, circuits.name AS circuits_name, results.milliseconds AS milliseconds FROM results 
JOIN drivers ON drivers.driverid = results.driverid AND drivers.dob > '1980-01-01' AND results.position = 1
JOIN races ON races.raceid = results.raceid
JOIN circuits ON circuits.circuitid = races.circuitid
ORDER BY drivers.dob DESC;

21 rows affected.


[('Sort  (cost=787.46..787.63 rows=69 width=62) (actual time=34.296..35.085 rows=199 loops=1)',),
 ('  Sort Key: drivers.dob DESC',),
 ('  Sort Method: quicksort  Memory: 52kB',),
 ('  ->  Nested Loop  (cost=25.68..785.35 rows=69 width=62) (actual time=1.808..32.823 rows=199 loops=1)',),
 ('        ->  Nested Loop  (cost=25.54..773.21 rows=69 width=50) (actual time=1.777..27.012 rows=199 loops=1)',),
 ('              ->  Hash Join  (cost=25.26..746.51 rows=69 width=27) (actual time=1.726..21.696 rows=199 loops=1)',),
 ('                    Hash Cond: (results.driverid = drivers.driverid)',),
 ('                    ->  Seq Scan on results  (cost=0.00..708.96 rows=974 width=24) (actual time=0.019..12.451 rows=974 loops=1)',),
 ('                          Filter: ("position" = 1)',),
 ('                          Rows Removed by Filter: 22703',),
 ('                    ->  Hash  (cost=24.51..24.51 rows=60 width=19) (actual time=1.677..1.677 rows=61 loops=1)',),
 ('                          Buckets: 1024  Batches: 1  Memory Usage: 12kB',),
 ('                          ->  Seq Scan on drivers  (cost=0.00..24.51 rows=60 width=19) (actual time=0.013..1.114 rows=61 loops=1)',),
 ("                                Filter: (dob > '1980-01-01'::date)",),
 ('                                Rows Removed by Filter: 780',),
 ('              ->  Index Scan using idx_17199_primary on races  (cost=0.28..0.39 rows=1 width=39) (actual time=0.009..0.009 rows=1 loops=199)',),
 ('                    Index Cond: (raceid = results.raceid)',),
 ('        ->  Index Scan using idx_17102_primary on circuits  (cost=0.14..0.18 rows=1 width=28) (actual time=0.011..0.011 rows=1 loops=199)',),
 ('              Index Cond: (circuitid = races.circuitid)',),
 ('Planning time: 0.721 ms',),
 ('Execution time: 35.941 ms',)]

As we can see from the result above then postgress is doing alot of things.
Postgres calculated that it took 0.721ms to figure out how to do this search properly and it actually took 35.941ms.
What we furthermore can see is that how postgres is doing it. I do not know exactly where is starts but i can see that we are ruuning multiple loops, filterings and sorting, i will walkthropigh it from the top. There are alot of info given in the table given above i will just make a quick walkthrough, if you want to see how long, loops and rows it did you will have to look in the table.
1. First we are doing a sort i would believe that the sort criteria is the Sort Key which is drivers.dob and that we are usings a quicksort.
2. The we are running two nested loops and a Hash Join with the criteria of a Hash Cond which is driverid.
3. Then we a re running a Seq scan to filter on position = 1 here we see that is removes 22703 rows.
4. Running a Hash to store the result.
5. Running a Seq scan with the filter og dob where the consdition is 1980-01-01 and removes 780 rows.
6. Running a Index Scan to join races with results.
7. Running a Index Scan to join circuits with races. 

# 4. Create a materialized view of your query. Using EXPLAIN ANALYZE try to query the view. Write at least 5 lines of text explaining what's going on and why the query execution time changed.

In [39]:
%%sql CREATE MATERIALIZED VIEW race_winners_cache AS
SELECT drivers.surname AS driver_surname, drivers.dob AS driver_age, races.name AS race_name, races.date AS race_date, circuits.name AS circuits_name, results.milliseconds AS milliseconds  FROM results 
JOIN drivers ON drivers.driverid = results.driverid AND drivers.dob > '1980-01-01' AND results.position = 1
JOIN races ON races.raceid = results.raceid
JOIN circuits ON circuits.circuitid = races.circuitid
ORDER BY drivers.dob DESC;

199 rows affected.


[]

In [40]:
%sql EXPLAIN ANALYZE SELECT * FROM race_winners_cache;

3 rows affected.


[('Seq Scan on race_winners_cache  (cost=0.00..10.50 rows=50 width=1564) (actual time=0.018..1.617 rows=199 loops=1)',),
 ('Planning time: 0.126 ms',),
 ('Execution time: 3.312 ms',)]

As we can see from this result then the time to run the query is a lot faster. In this Select statement we are only running one Seq Scan to get everyting from that view. But what we need to remember is that this view only contains the data we gave is at that time so it is not updated. That is also why this is alot faster i believe. If we should update this we would need to run the query agian, which still takes aroung 30ms to exceute. But this is a great solution if the data that we have created the view on is not changed so much.